File: day02.hs
import Data.Bifunctor (bimap)

data Play = Rock | Paper | Scissors
data Result = Win | Lose | Tie

firstParse :: String -> Play
firstParse s = case s of
  "X" -> Rock
  "Y" -> Paper
  "Z" -> Scissors
  _   -> error "Invalid Input"

secondParse :: String -> Result
secondParse s = case s of
  "X" -> Lose
  "Y" -> Tie
  "Z" -> Win
  _   -> error "Invalid Input"
  
scorePair :: (Play, Play) -> Int
scorePair  p = case p of
  (Rock, Rock) -> 1 + 3
  (Rock, Paper) -> 2 + 6
  (Rock, Scissors) -> 3 + 0
  (Paper, Rock) ->  1 + 0
  (Paper, Paper) -> 2 + 3
  (Paper, Scissors) -> 3 + 6
  (Scissors, Rock) -> 1 + 6
  (Scissors, Paper) -> 2 + 0
  (Scissors, Scissors) -> 3 + 3 
    
partOne :: [(Play, Play)] -> Int
partOne = sum . map scorePair 
        
partTwo :: [(Play, Result)] -> Int
partTwo = sum . map (scorePair . resToPlay)
  where
    -- This part may seem odd, because (Rock, Win) reads like "How do I win with Rock" but it's
    -- actually 'How do I win against Rock'
    resToPlay :: (Play, Result) -> (Play, Play)
    resToPlay p = case p of
      (Rock, Win) -> (Rock, Paper)
      (Rock, Tie) -> (Rock, Rock)
      (Rock, Lose) -> (Rock, Scissors)
      (Paper, Win) -> (Paper, Scissors)
      (Paper, Tie) -> (Paper, Paper)
      (Paper, Lose) -> (Paper, Rock)
      (Scissors, Win) -> (Scissors, Rock)
      (Scissors, Tie) -> (Scissors, Scissors)
      (Scissors, Lose) -> (Scissors, Paper)

clean :: String -> (String -> b) -> [(Play, b)]
clean s toB = map toPair . lines $ s
  where
    toPair line = case line of
      ('A':_:me) -> (Rock, toB me)
      ('B':_:me) -> (Paper, toB me)
      ('C':_:me) -> (Scissors, toB me)
      _          -> error $ "Malformed input: '" ++ line ++ "'"

main :: IO ()
main  = do
  input <- readFile "./input/day02.txt"
  let cleanInput = clean input in do
    print . partOne $ cleanInput firstParse
    print . partTwo $ cleanInput secondParse
    

File: day02.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum {
  Rock = 'X',
  Paper = 'Y',
  Scissors = 'Z',

  Loss = 'X',
  Tie = 'Y',
  Win = 'Z',
};

long long score_game(int opp, int me) {
  if ((opp == Rock && me == Paper) || (opp == Scissors && me == Rock) || (opp == Paper && me == Scissors)) {
    return 6;
  } else if ((opp == Rock && me == Scissors) || (opp == Scissors && me == Paper) || (opp == Paper && me == Rock)) {
    return 0;
  }  else {
    return 3;
  }
}

int get_play(int opp, int result) {
  if ((opp == Rock && result == Loss) || (opp == Paper && result == Win)) {
    return Scissors;
  } else if ((opp == Scissors && result == Loss) || (opp == Rock && result == Win)) {
    return Paper;
  } else if ((opp == Paper && result == Loss) || (opp == Scissors && result == Win)) {
    return Rock;
  } else {
    return opp;
  }
}

void part_one(char const* input, long length) {
  long long score = 0;
  
  for(int x = 0; x < length; x += 4) {
    char opp  = input[x] + 23;
    char me   = input[x + 2];

    score += score_game(opp, me);

    switch (me) {
      case Rock:     score += 1; break;
      case Paper:    score += 2; break;           
      case Scissors: score += 3; break;            
    }
  }
  
  printf("Score for part 1: %lld\n", score);
} 

void part_two(char const* input, long length) {
  long long score = 0;
  
  for(int x = 0; x < length; x += 4) {
    char opp    = input[x] + 23;
    char result = input[x + 2];
    char me     = get_play(opp, result);

    score += score_game(opp, me);

    switch (me) {
      case Rock:     score += 1; break;
      case Paper:    score += 2; break;           
      case Scissors: score += 3; break;            
    }
  }
  
  printf("Score for part 1: %lld\n", score);
}

int main(int argc, char* argv[]) {
  FILE* file = fopen("./input/day02.txt", "r");
  if (file == NULL) {
    fprintf(stderr, "Unable to open file...");
    return 1;
  }
  
  fseek(file, 0, SEEK_END);
  long length = ftell(file);
  fseek(file, 0, SEEK_SET);
  
  char* string = malloc(length + 1);
  if (string == NULL) {
    fprintf(stderr, "Malloc failed... ");
    return 1;
  } 
  
  fread(string, 1, length, file);
  string[length] = '\n';

  fclose(file);
     
  part_one(string, length);
  part_two(string, length);
}
